File: design_pattern/chain_0f_responsibility/logger.java
public enum LogLevel {
    INFO,
    DEBUG,
    WARN,
    ERROR
};

public interface Logger{
    void log(LogLevel level, String message);

}

public abstract class BaseLogger implements Logger{
    protected Logger nextLogger = null;
    public BaseLogger(Logger nextLogger) {
        this.nextLogger = nextLogger;
    }
}

public class ConsoleLogger extends BaseLogger {

    public ConsoleLogger(Logger nextLogger) {
        super(nextLogger);
    }

    @Override
    public void log(LogLevel level, String message) {
        if (level == LogLevel.INFO || level == LogLevel.DEBUG) {
            System.out.println("Console Logger: " + message);
        }
        if (nextLogger != null) {
            nextLogger.log(level, message);
        }
    }
}

public class WarnLogger extends BaseLogger {

    public FileLogger(Logger nextLogger) {
        super(nextLogger);
    }

    @Override
    public void log(LogLevel level, String message) {
        if (level == LogLevel.WARN) {
            System.out.println("File Logger: " + message);
        }
        if (nextLogger != null) {
            nextLogger.log(level, message);
        }
    }
}

public class ErrorLogger extends BaseLogger {

    public ErrorLogger(Logger nextLogger) {
        super(nextLogger);
    }

    @Override
    public void log(LogLevel level, String message) {
        if (level == LogLevel.ERROR) {
            System.out.println("Error Logger: " + message);
        }
        if (nextLogger != null) {
            nextLogger.log(level, message);
        }
    }
}

public class LoggerFactory {
    public static Logger createLogger() {
        Logger errorLogger = new ErrorLogger(null);
        Logger warnLogger = new WarnLogger(errorLogger);
        Logger consoleLogger = new ConsoleLogger(warnLogger);
        return consoleLogger;
    }
}

public class LoggerDemo {
    public static void main(String[] args) {
        Logger logger = LoggerFactory.createLogger();

        logger.log(LogLevel.INFO, "This is an info message.");
        logger.log(LogLevel.DEBUG, "This is a debug message.");
        logger.log(LogLevel.WARN, "This is a warning message.");
        logger.log(LogLevel.ERROR, "This is an error message.");
    }
}

File: design_pattern/chain_0f_responsibility/logger.go
package main

import (
	"fmt"
)

// Enum-like log level
type LoggerType int

const (
	DEBUG LoggerType = iota
	INFO
	WARN
	ERROR
)

// Logger interface defines the methods
type Logger interface {
	SetNext(next Logger)
	Log(level LoggerType, message string)
}

// Base struct to share next-handler logic
type BaseLogger struct {
	next Logger
}

func (b *BaseLogger) SetNext(next Logger) {
	b.next = next
}

func (b *BaseLogger) CallNext(level LoggerType, message string) {
	if b.next != nil {
		b.next.Log(level, message)
	}
}

// DebugLogger
type DebugLogger struct {
	BaseLogger
}

func (d *DebugLogger) Log(level LoggerType, message string) {
	if level == DEBUG {
		fmt.Println("Debug:", message)
	}
	d.CallNext(level, message)
}

// InfoLogger
type InfoLogger struct {
	BaseLogger
}

func (i *InfoLogger) Log(level LoggerType, message string) {
	if level == INFO {
		fmt.Println("Info:", message)
	}
	i.CallNext(level, message)
}

// WarnLogger
type WarnLogger struct {
	BaseLogger
}

func (w *WarnLogger) Log(level LoggerType, message string) {
	if level == WARN {
		fmt.Println("Warn:", message)
	}
	w.CallNext(level, message)
}

// ErrorLogger
type ErrorLogger struct {
	BaseLogger
}

func (e *ErrorLogger) Log(level LoggerType, message string) {
	if level == ERROR {
		fmt.Println("Error:", message)
	}
	e.CallNext(level, message)
}

func main() {
	// Construct chain: Debug -> Info -> Warn -> Error
	debugLogger := &DebugLogger{}
	infoLogger := &InfoLogger{}
	warnLogger := &WarnLogger{}
	errorLogger := &ErrorLogger{}

	debugLogger.SetNext(infoLogger)
	infoLogger.SetNext(warnLogger)
	warnLogger.SetNext(errorLogger)

	// Start logging through the chain
	debugLogger.Log(DEBUG, "This is a debug message")
	debugLogger.Log(INFO, "This is an info message")
	debugLogger.Log(WARN, "This is a warning message")
	debugLogger.Log(ERROR, "This is an error message")
}
